cosmwasm/enclaves/shared/contract-engine/src/wasm/contract.rs

```diff
@@ -603,4 +654,677 @@ impl WasmiApi for ContractInstance {
+            debug!("ed25519_batch_verify() error while trying to read signatures from wasm memory");
+            err
+        })?;
+
+        let pubkeys_data = self
+            .decode_sections(public_keys_ptr as u32)
+            .map_err(|err| {
+                debug!(
+                    "ed25519_batch_verify() error while trying to read public_keys from wasm memory"
+                );
+                err
+            })?;
+
+        let (messages, signatures, pubkeys) = if messages_data.len() == signatures_data.len()
+            && messages_data.len() == pubkeys_data.len()
+        {
+            // All is well, convert to Vec<&[u8]>
+            (
+                messages_data
+                    .iter()
+                    .map(|m| m.as_slice())
+                    .collect::<Vec<&[u8]>>(),
+                signatures_data
+                    .iter()
+                    .map(|s| s.as_slice())
+                    .collect::<Vec<&[u8]>>(),
+                pubkeys_data
+                    .iter()
+                    .map(|p| p.as_slice())
+                    .collect::<Vec<&[u8]>>(),
+            )
+        } else if messages_data.len() == 1 && signatures_data.len() == pubkeys_data.len() {
+            // Multisig, replicate message
+            (
+                vec![messages_data[0].as_slice()].repeat(signatures_data.len()),
+                signatures_data
+                    .iter()
+                    .map(|s| s.as_slice())
+                    .collect::<Vec<&[u8]>>(),
+                pubkeys_data
+                    .iter()
+                    .map(|p| p.as_slice())
+                    .collect::<Vec<&[u8]>>(),
+            )
+        } else if pubkeys_data.len() == 1 && messages_data.len() == signatures_data.len() {
+            // Replicate pubkey
+            (
+                messages_data
+                    .iter()
+                    .map(|m| m.as_slice())
+                    .collect::<Vec<&[u8]>>(),
+                signatures_data
+                    .iter()
+                    .map(|s| s.as_slice())
+                    .collect::<Vec<&[u8]>>(),
+                vec![pubkeys_data[0].as_slice()].repeat(signatures_data.len()),
+            )
+        } else {
+            debug!(
+                "ed25519_batch_verify() mismatched number of messages ({}) / signatures ({}) / public keys ({})", messages_data.len(),signatures_data.len(),pubkeys_data.len()
+            );
+
+            // return 5 == BatchErr
+            // https://github.com/CosmWasm/cosmwasm/blob/v1.0.0-beta5/packages/crypto/src/errors.rs#L97
+            return Ok(Some(RuntimeValue::I32(7)));
+        };
+
+        self.use_gas_externally(
+            (self.gas_costs.external_ed25519_verify
+                - self.gas_costs.external_ed25519_batch_verify_per_one) as u64 /* base cost in case signatures.len() == 0 */
+                + (signatures.len() as u64) * self.gas_costs.external_ed25519_batch_verify_per_one as u64,
```

Comment:
I don't really understand this calculation. Why `self.gas_costs.external_ed25519_verify - self.gas_costs.external_ed25519_batch_verify_per_one`?  (also `per_one` should be either `single` or `each`)

Reply:
To prevent a scenario of zero gas if the caller loops over this API with `signatures.len() == 0`.

Reply:
So define a new self.gas_costs.external_ed25519_batch_base_cost or something? Having it like this will cause unintended cosequences if someone tries to set verify < batch_verify_per_one